w3m and ranger fixes

oxij opened the pull request:
Rebase of #10995 onto staging

oxij committed:
w3m: fix w3mimgdisplay, refactor the expression, make batch and nox versions, use batch version where appropriate

jagajaga commented:
cc @edolstra

garbas commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13387,8 +13390,12 @@ let
+  w3m-pure = w3m.override {
     graphicsSupport = false;
+    x11Support = false;
+    mouseSupport = false;
```
@oxij awesome work. 

i have only once comment just to keep things consistent with other derivations. usually we separate a package into <name> and <name>Full ... example (python and pythonFull). i think here we should do the same and have

w3m be what currently is w3m-pure and create new w3mFull which would bring all the graphicSupport, x11Support, mouseSupport. 

basically all i would like is to things be more or less consistent. @oxij you should wait with any changes until somebody else comments and confirms this.

oxij replied:
Yep, it is more consistent, consistent is good, I can do that. But even
then I'd still like to have an explicitly minimal version in case
someone else will fiddle with defaults in the future. `w3m-pure` is for
derivations that use w3m in batch mode for text rendering, not browsing,
and I'd like that information to be explicit in nixpkgs. Which means
there would be three `w3m`'s: `w3m`, `w3mFull`, `w3mMinimal` (current
`w3m-pure`) and

* I have no idea what should be the difference between `w3m` and
  `w3mFull` if the default derivation should be what "most users will want
  by default".
* ranger will have to use `w3mFull` or specifically crafted version (4th
  `w3m`!) with image support.

So, should I rename `w3m-pure` to `w3mMinimal` and define `w3mFull =
w3m`?

(Btw, I still think that using nixpkgs.config options (constantly
rejected in nixpkgs) is a good thing as it simplifies things a lot, e.g.
it can simplify these four `w3m`s into two and one assert in `ranger`,
but whatever.)

nbp commented on pkgs/top-level/all-packages.nix:
```diff
@@ -13387,8 +13390,12 @@ let
-  w3m = callPackage ../applications/networking/browsers/w3m {
+  w3m = callPackage ../applications/networking/browsers/w3m { };
+
+  w3m-pure = w3m.override {
```
Use `self.w3m.override`, otherwise this would add a security issue when we fix `w3m`.

oxij replied:
@nbp eh, why? AFAICS everything in all-packages.nix overrides stuff this way

nbp replied:
See https://github.com/NixOS/nixpkgs/pull/10851#discussion_r45405883 for a detailed explanation.

vcunat commented:
I picked the ranger commit, as that one certainly doesn't need to go through staging. And it looks fine, thanks.

cstrahan commented on pkgs/applications/networking/browsers/w3m/default.nix:
```diff
@@ -40,7 +43,12 @@ stdenv.mkDerivation rec {
-  meta = with stdenv.lib; {
+  postInstall = optionalString x11Support ''
+    rpath=`patchelf --print-rpath $out/libexec/w3m/w3mimgdisplay`
+    patchelf --set-rpath "$rpath:${libX11}/lib:${libX11}/lib64" $out/libexec/w3m/w3mimgdisplay
```
You can avoid this by using:

```
  LIBS = lib.optionalString graphicsSupport "-lX11";
```

The autoconf stuff will pick up that environment variable.

cstrahan commented on pkgs/applications/networking/browsers/w3m/default.nix:
```diff
@@ -1,16 +1,18 @@
+, ncurses, boehmgc, gettext, zlib
+, sslSupport ? true, openssl ? null
+, graphicsSupport ? true, imlib2 ? null
+, x11Support ? graphicsSupport, libX11 ? null
```
I'm pretty sure that `graphicsSupport` should imply `x11Support`, so I don't think we should add this option (unless I'm mistaken).

oxij replied:
`LIB` thingy is okay, yep, I'll fix that.

`graphicsSupport` thingy in your comment and PR is not okay, because my
patch also works without X11 in frame buffer (I actually use that with
`ranger`).

I also don't like the idea of `w3m == w3m-pure` because then `ranger`,
`emacs` and everything else that uses `w3m` as a browser (that is, all
user-oriented stuff) will need to be overriden which is not nice. I can
agree to rename `w3m-pure` to `w3mBatch` and define `w3mFull = w3m` if
that bothers anyone, but nothing less.

I say after I fix `LIB` thingy, test that it doesn't break the frame
buffer and everyone agrees to these changes, then you should apply your
`fetchpatch`-changes (on top or in parallel, shouldn't matter, to this).

oxij replied:
> I say after I fix `LIB` thingy, test that it doesn't break the frame
> buffer and everyone agrees to these changes, then you should apply your
> `fetchpatch`-changes (on top or in parallel, shouldn't matter, to this).

... because this way it would be much easier to merge for both of us:
this way you'll need to throw away changes from your patches and it will
just work, while the way proposed by you'll require merging at the top
of the file.

cstrahan replied:
> `graphicsSupport` thingy in your comment and PR is not okay, because my
patch also works without X11 in frame buffer (I actually use that with
`ranger`).

Ah, OK. I thought `imlib` was used only for rendering, but if it serves another text-only purpose, then I agree with your rationale.

> I also don't like the idea of `w3m == w3m-pure` because then `ranger`,
`emacs` and everything else that uses `w3m` as a browser (that is, all
user-oriented stuff) will need to be overriden which is not nice. I can
agree to rename `w3m-pure` to `w3mBatch` and define `w3mFull = w3m` if
that bothers anyone, but nothing less.

Yeah, that is an interesting point. I could see using `w3mPure` where needed, then one could override `w3m` to enable whatever they want, and then that would get picked up in reverse dependencies. Sounds good.

> I say after I fix `LIB` thingy, test that it doesn't break the frame
buffer and everyone agrees to these changes, then you should apply your
`fetchpatch`-changes (on top or in parallel, shouldn't matter, to this).

Sounds good.

Thanks for this PR :beer:

oxij replied:
Okay then, rebuilding the thing right now (will take a while with `gcc` update).

I decided to do this:

   w3m = callPackage ../applications/networking/browsers/w3m { };

   # Should always be the version with the most features
   w3m-full = w3m;
 
   # Version without X11
   w3m-nox = w3m.override {
     x11Support = false;
   };
 
   # Version for batch text processing, not a good browser
   w3m-batch = w3m.override {
     graphicsSupport = false;
     x11Support = false;
     mouseSupport = false;
   };

Does this look good to you?

Thank you, too!

cstrahan replied:
That looks good to me.

luigy replied:
`w3m-batch`  is not darwin friendly since `w3m` refuses to evaluate because of `gpm` by having `mouseSupport` set to true by default


hmm I'm not too familiar with nixpkgs, but I'm currently using `callPackage` instead of `override` and I guess the same would be required for `w3m-nox`.
```
w3m-batch = callPackage ../applications/networking/browsers/w3m {
  graphicsSupport = false;
  x11Support = false;
  mouseSupport = false;
};
```

oxij replied:
Does the following help?


diff --git a/pkgs/applications/networking/browsers/w3m/default.nix b/pkgs/applications/networking/browsers/w3m/default.nix
index e71a733..bea74f6 100644
--- a/pkgs/applications/networking/browsers/w3m/default.nix
+++ b/pkgs/applications/networking/browsers/w3m/default.nix
@@ -3,7 +3,7 @@
 , sslSupport ? true, openssl ? null
 , graphicsSupport ? true, imlib2 ? null
 , x11Support ? graphicsSupport, libX11 ? null
-, mouseSupport ? true, gpm-ncurses ? null
+, mouseSupport ? !stdenv.isDarwin, gpm-ncurses ? null
 }:

luigy replied:
nice! that works for me

cstrahan commented:
I think my PR (https://github.com/NixOS/nixpkgs/pull/11236) subsumes much of this, with the exception of touching the names in all-packages.

Regarding the names, I'd suggest having `w3m` be minimal, and have `w3mFull` (or something along those lines) be w3m with all of the bells and whistles.

I would suggest pulling in my PR, and then layering on these changes. How does that sound?

oxij committed:
nixos: use w3m-nox for the manual

oxij commented:
Okay, I think I fixed everything except @nbp's comments (I don't see
this security thing being merged yet and I don't want to run in front of
the train here), actually checked this to work without X11 (which added
another patch :( ) and fixed nixos to use `w3m-nox` as it should.

oxij commented:
Just wondering, what is holding this?

magnetophon commented:
Great, I was just about to work out proper img-previewing in ranger; thanks!
Would it make sense to backport to 15.09 as well?

oxij commented:
> Would it make sense to backport to 15.09 as well?

IMHO, yes.

cstrahan commented:
> Great, I was just about to work out proper img-previewing in ranger; thanks!

I find it interesting that we were looking into this at roughly the same time :)
